### PR TITLE
Add support for ILIKE in the where clause

### PR DESCRIPTION
**Pull request by marregui**
Replacement for https://github.com/crate/crate/pull/9150 

## Checklist

 - [ ] User relevant changes are recorded in ``CHANGES.txt``
 - [ ] Touched code is covered by tests
 - [ ] Documentation has been updated if necessary
 - [ ] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [ ] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by marregui** [comment]
retest this please

flaky blackbox/docs/general/occ.rst I can't reproduce locally

File: sql/src/main/java/io/crate/lucene/LuceneQueryBuilder.java [hunk 1]
```diff
@@ -387,7 +386,8 @@ public Query apply(Function input, Context context) {
                 .put(LteOperator.NAME, LTE_QUERY)
                 .put(GteOperator.NAME, GTE_QUERY)
                 .put(GtOperator.NAME, GT_QUERY)
-                .put(LikeOperator.NAME, new LikeQuery())
+                .put(LikeOperators.OP_LIKE, new LikeQuery())
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/lucene/LuceneQueryBuilder.java]: Maybe we could pass in the ignore case parameter instead of extracting it inside the classes functions

File: sql/src/test/java/io/crate/analyze/GroupByAnalyzerTest.java [hunk 2]
```diff
@@ -344,13 +345,23 @@ public void testGroupWithInvalidNullLiteral() throws Exception {
...
-        assertThat(relation.having().query(), isFunction("op_like"));
+        assertThat(relation.having().query(), isFunction(LikeOperators.OP_LIKE));
         Function havingFunction = (Function) relation.having().query();
         assertThat(havingFunction.arguments().size(), is(2));
         assertThat(havingFunction.arguments().get(0), isReference("name"));
         assertThat(havingFunction.arguments().get(1), isLiteral("Slartibart%"));
     }
 
+    @Test
+    public void testGroupByHaving2() throws Exception {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/analyze/GroupByAnalyzerTest.java]: Can we have more verbose test names that hint at the concrete scenario being tested?

File: sql/src/test/java/io/crate/integrationtests/SubSelectIntegrationTest.java [hunk 3]
```diff
@@ -65,6 +65,13 @@ public void testSubSelectWhere() throws Exception {
                 "where name like 'Arthur'");
         assertThat(printedTable(response.rows()),
             is("4| Arthur\n"));
+
+        execute("select id, name " +
+                "from (select * from characters where female = true) as ch " +
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/SubSelectIntegrationTest.java]: please don't add a ilike test into a test case for something else. This makes changing tests later on more difficult as it will be more unclear what the test was supposed to test

We can use the QueryTester framework to very that the built Lucene query behave the way we want.

File: sql/src/main/java/io/crate/expression/symbol/format/SymbolPrinter.java [hunk 4]
```diff
@@ -187,7 +192,10 @@ private void printAnyOperator(Function function, SymbolPrinterContext context) {
             args.get(0).accept(this, context);
 
             // print operator
-            String operatorName = anyOperatorName(function.info().ident().name());
+            boolean ignoreCase = LikeOperators.extractIgnoreCase(function);
+            String functionName = function.info().ident().name();
+            String operatorName = caseSensitiveOperatorName(anyOperatorName(functionName), ignoreCase);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/expression/symbol/format/SymbolPrinter.java]: Is this still required now that we have separate ilike operator functions?

File: sql/src/main/java/io/crate/lucene/AbstractAnyQuery.java [hunk 5]
```diff
@@ -70,16 +80,18 @@ public Query apply(Function function, LuceneQueryBuilder.Context context) throws
      * converts Strings to BytesRef on the fly
      */
     static Iterable<?> toIterable(Object value) {
-        return Iterables.transform(AnyOperators.collectionValueToIterable(value), new com.google.common.base.Function<Object, Object>() {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/lucene/AbstractAnyQuery.java]: please avoid unrelated formatting changes like this.

File: sql/src/test/java/io/crate/analyze/GroupByAnalyzerTest.java [hunk 6]
```diff
@@ -344,13 +345,23 @@ public void testGroupWithInvalidNullLiteral() throws Exception {
...
-        assertThat(relation.having().query(), isFunction("op_like"));
+        assertThat(relation.having().query(), isFunction(LikeOperators.OP_LIKE));
         Function havingFunction = (Function) relation.having().query();
         assertThat(havingFunction.arguments().size(), is(2));
         assertThat(havingFunction.arguments().get(0), isReference("name"));
         assertThat(havingFunction.arguments().get(1), isLiteral("Slartibart%"));
     }
 
+    @Test
+    public void testCaseInsensitiveGroupByHaving() throws Exception {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/analyze/GroupByAnalyzerTest.java]: I think the test names could still be further improved. E.g. `test_ilike_can_be_used_in_having_clause` (which makes me wonder why this is a special case and why we'd test it?)

We're not testing all operator placement permutations.

File: sql/src/test/java/io/crate/analyze/GroupByAnalyzerTest.java [hunk 7]
```diff
@@ -420,6 +442,20 @@ public void testGroupByHavingComplex() throws Exception {
         assertThat(andFunction.arguments().get(1), isFunction("op_not"));
     }
 
+    @Test
+    public void testGroupByHavingComplex2() throws Exception {
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/analyze/GroupByAnalyzerTest.java]: Same thing with these tests here.. `Complex2` has no indication of how this is different from `testGroupByHavingComplex` (which is also a bad test name to begin with).

Keep in mind for test cases: "Could I re-create the test body given the test-case name?" 

If not it is really difficult to reason if a test is still required after a refactoring that would affect this test somehow.

File: sql/src/test/java/io/crate/integrationtests/SubSelectIntegrationTest.java [hunk 8]
```diff
@@ -65,6 +65,21 @@ public void testSubSelectWhere() throws Exception {
                 "where name like 'Arthur'");
         assertThat(printedTable(response.rows()),
             is("4| Arthur\n"));
+
+        assertThat(printedTable(response.rows()),
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/integrationtests/SubSelectIntegrationTest.java]: This is redundant?

File: sql/src/main/java/io/crate/expression/symbol/format/SymbolPrinter.java (hunk 4 above)
Comment (marregui) [reply sql/src/main/java/io/crate/expression/symbol/format/SymbolPrinter.java]: No, good eye!, thank you

**Comment by marregui** [comment]
Regarding the tests, indeed, I have removed those that were unnecessary, I have restored those that mixed ilike/like and I have kept those that I think are relevant. As for the naming, yes, going forward I will use longer names, keeping in mind that I should be able to reconstruct their bodies from the name alone. I will use the underscore. For the tests that I have added, I have followed the style of the class file, with the camel case.

File: sql/src/main/java/io/crate/lucene/LikeQuery.java [hunk 9]
```diff
@@ -39,37 +41,60 @@
...
+
+    private static Query createCaseAwareQuery(String fieldName,
+                                              String text,
+                                              boolean ignoreCase) {
+        java.util.function.Function<String, String> regexTransformer = ignoreCase ?
+            LikeOperators::patternToRegex
+            :
+            LikeQuery::convertSqlLikeToLuceneWildcard;
+        Term term = new Term(fieldName, regexTransformer.apply(text));
+        return ignoreCase ? new CrateRegexQuery(term, CASE_INSENSITIVE) : new WildcardQuery(term);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/lucene/LikeQuery.java]: Let's follow up by adding/extending some benchmarks for ilike vs. like in https://github.com/crate/crate-benchmarks
